### ISSUE TITLE
Bugs in training shadow models when the shadow model template is a PyTorchClassifier 

### ISSUE THREAD

**Issue by EmmyFang** [Trusted-AI/adversarial-robustness-toolbox#1732]
**Describe the bug**
The shadow models don't get trained when the model template is a PyTorch model. (`from art.attacks.inference.membership_inference import ShadowModels`). The training accuracy of the shadow models is really low. 

**To Reproduce**
``` 
import os
import sys
sys.path.insert(0, os.path.abspath('..'))

import numpy as np

from art.utils import load_nursery

(x_target, y_target), (x_shadow, y_shadow), _, _ = load_nursery(test_set=0.75)

target_train_size = len(x_target) // 2
x_target_train = x_target[:target_train_size]
y_target_train = y_target[:target_train_size]
x_target_test = x_target[target_train_size:]
y_target_test = y_target[target_train_size:]

import torch
from torch import nn, optim
from torch.utils.data import DataLoader
from torch.utils.data.dataset import Dataset
from art.estimators.classification.pytorch import PyTorchClassifier
import numpy as np

class ModelToAttack(nn.Module):

    def __init__(self, num_classes, num_features):
        super(ModelToAttack, self).__init__()

        self.fc1 = nn.Sequential(
                nn.Linear(num_features, 128),
                nn.ReLU(), )

        
        self.classifier = nn.Linear(128, num_classes)

    def forward(self, x):
        out = self.fc1(x)
        return self.classifier(out)

from art.attacks.inference.membership_inference import ShadowModels
from art.utils import to_categorical

mlp_model = ModelToAttack(4, 24)
# mlp_model = torch.nn.DataParallel(mlp_model)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(mlp_model.parameters(), lr=0.01)

mlp_art_model = PyTorchClassifier(model=mlp_model, loss=criterion, optimizer=optimizer, input_shape=(24,), nb_classes=4)

mlp_art_model.fit(x_target_train.astype(np.float32), y_target_train)
pred = np.array([np.argmax(arr) for arr in mlp_art_model.predict(x_target_test.astype(np.float32))])
print('Base model accuracy: ', np.sum(pred == y_target_test) / len(y_target_test))


shadow_models = ShadowModels(mlp_art_model, num_shadow_models=3)

shadow_dataset = shadow_models.generate_shadow_dataset(x_shadow, to_categorical(y_shadow, 4))
(member_x, member_y, member_predictions), (nonmember_x, nonmember_y, nonmember_predictions) = shadow_dataset

# Shadow models' accuracy
# print([sm.model.score(x_target_test, y_target_test) for sm in shadow_models.get_shadow_models()])
for sm in shadow_models.get_shadow_models():
    pred = np.array([np.argmax(arr) for arr in sm.predict(x_target_test.astype(np.float32), 100, False)])
    print('shadow model accuracy: ', np.sum(pred == y_target_test) / len(y_target_test))

for sm, sm_train in zip(shadow_models.get_shadow_models(), shadow_models.get_shadow_models_train_sets()):
    pred = np.array([np.argmax(arr) for arr in sm.predict(sm_train[0], 100, False)])
    print('shadow model train accuracy: ', np.sum(pred == np.argmax(sm_train[1], 1)) / len(np.argmax(sm_train[1], 1)))
```

**Expected behavior**
I got the following outputs: 
```
Base model accuracy:  0.9660493827160493
shadow model accuracy:  0.21358024691358024
shadow model accuracy:  0.15987654320987654
shadow model accuracy:  0.1

shadow model train accuracy:  0.2198888202594194
shadow model train accuracy:  0.16553428042001236
shadow model train accuracy:  0.10315009264978382
```
The base model reaches reasonable test accuracy. However, the test accuracies of the shadow models are very low. I tested with their shadow model train sets, the training accuracy is also really low. It seems the shadow models didn't get trained. 

**System information:**
 - OS: MacOS 12.3.1
 - Python version: 3.6.13
 - ART version or commit number: version 1.10.2
 - TensorFlow / Keras / PyTorch / MXNet version: PyTorch 1.10

**Comment by beat-buesser** [comment]
Hi @EmmyFang Thank you very much for raising this issue. We'll take a closer look as soon as possible.

**Comment by beat-buesser** [comment]
Hi @EmmyFang The solution shared by @chao1995 in #1742 fixes this issue.
